File: ios/Bill24SdkModule.swift
import ExpoModulesCore
import B24PaymentSdk

public class Bill24SdkModule: Module {

  public func definition() -> ModuleDefinition {

    Name("Bill24Sdk")

    AsyncFunction("initializePayment") {
            (param: Bill24InstantPaymentSdk) -> [String: Any] in

            if param.userSyncCode.isEmpty {
                return [
                    "message": "userSyncCode is required",
                    "success": false,
                ]
            }
            if param.refererKey.isEmpty {
                return [
                    "message": "refererKey is required",
                    "success": false,
                ]
            }

            guard
                let presentingViewController = self.appContext?.utilities?
                    .currentViewController()
            else {

                return [
                    "message": "failed to get current view controller",
                    "success": false,
                ]

            }

            B24PaymentSdk.instantPaymentSdk(
                from: presentingViewController,
                userSynCode: param.userSyncCode,
                refererKey: param.refererKey,
                language: param.language,
                isDarkMode: param.isDarkMode,
                isProduction: param.isProduction,
                testingEnv: param.testingEnv,
                isDisplayCurrencySymbol: param.isDisplayCurrencySymbol
            )

            return [
                "message": "Instant Payment SDK initialized successfully",
                "success": true,
            ]

        }.runOnQueue(.main)

        AsyncFunction("instantPaymentSdkTopUp") {
            (param: BIll24InstantPaymentTopUp) -> [String: Any] in
            if param.paymentMethodId.isEmpty {
                return [
                    "message": "paymentMethodId is required",
                    "success": false,
                ]
            }
            if param.refererKey.isEmpty {
                return [
                    "message": "refererKey is required",
                    "success": false,
                ]
            }
            guard
                let presentingViewController = self.appContext?.utilities?
                    .currentViewController()
            else {
                return [
                    "message": "failed to get current view controller",
                    "success": false,
                ]
            }
            B24PaymentSdk.instantPaymentSdkTopUp(
                from: presentingViewController,
                paymentMethodId: param.paymentMethodId,
                refererKey: param.refererKey,
                language: param.language,
                isDarkMode: param.isDarkMode,
                isProduction: param.isProduction,
                testingEnv: param.testingEnv,
                isDisplayCurrencySymbol: param.isDisplayCurrencySymbol
            )

            return [
                "message": "Instant Payment SDK Top Up initialized successfully",
                "success": true,
            ]
        }.runOnQueue(.main)

    }
}

File: android/src/main/java/com/rn/sdk/bill24/Bill24SdkModule.kt
package com.rn.sdk.bill24

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Queues
import com.bill24.b24paymentsdk.main.B24PaymentSdk
import com.rn.sdk.bill24.records.Bill24InstantPaymentSdk
import com.rn.sdk.bill24.records.Bill24InstantPaymentTopUp
import com.rn.sdk.bill24.records.Bill24PaymentResult

class Bill24SdkModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('Bill24Sdk')` in JavaScript.
    Name("Bill24Sdk")

    AsyncFunction("instantPaymentSdk") { params: Bill24InstantPaymentSdk ->
            if (params.userSyncCode.isBlank()) {
                return@AsyncFunction Bill24PaymentResult(
                    "userSyncCode is required", false
                )
            }
            if (params.refererKey.isBlank()) {
                return@AsyncFunction Bill24PaymentResult("refererKey is required", false)
            }
             B24PaymentSdk.instantPaymentSdk(
                 context,
                 params.userSyncCode,
                 params.refererKey,
                 params.language,
                 params.isDarkMode,
                 params.isProduction,
                 params.testingEnv,
                 params.isDisplayCurrencySymbol
             )
            return@AsyncFunction Bill24PaymentResult(
                "Instant Payment SDK initialized successfully", true
            )
        }.runOnQueue(Queues.MAIN)

        AsyncFunction("instantPaymentSdkTopUp") { params: Bill24InstantPaymentTopUp ->
            if (params.paymentMethodId.isBlank()) {
                return@AsyncFunction Bill24PaymentResult(
                    "paymentMethodId is required", false
                )
            }
            if (params.refererKey.isBlank()) {
                return@AsyncFunction Bill24PaymentResult(
                    "refererKey is required", false
                )
            }
             B24PaymentSdk.instantPaymentSdkTopUp(
                 currentActivity, // Use activity context instead
                 params.paymentMethodId,
                 params.refererKey,
                 params.language,
                 params.isDarkMode,
                 params.isProduction,
                 params.testingEnv,
                 params.isDisplayCurrencySymbol
             )
            return@AsyncFunction Bill24PaymentResult(
                "Instant Payment SDK Top Up initialized successfully", true
            )
        }.runOnQueue(Queues.MAIN)
    }

    private val context: Context
        get() = requireNotNull(appContext.reactContext) { "React Application Context is null" }

    private val currentActivity
        get() = appContext.currentActivity ?: throw Exceptions.MissingActivity()
}
